x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java

```diff
@@ -305,8 +308,16 @@ static IdpConfiguration getIdpConfiguration(
         final PredicateRoleDescriptorResolver roleDescriptorResolver = new PredicateRoleDescriptorResolver(metadataResolver);
         resolver.setRoleDescriptorResolver(roleDescriptorResolver);
 
-        final InlineX509DataProvider keyInfoProvider = new InlineX509DataProvider();
-        resolver.setKeyInfoCredentialResolver(new BasicProviderKeyInfoCredentialResolver(Collections.singletonList(keyInfoProvider)));
+        final List<KeyInfoProvider> keyInfoProviders = Collections.singletonList(new InlineX509DataProvider());
+        final BasicProviderKeyInfoCredentialResolver credentialsResolver = new BasicProviderKeyInfoCredentialResolver(keyInfoProviders) {
+            @Override
+            protected void postProcess(KeyInfoResolutionContext kiContext, CriteriaSet criteriaSet, List<Credential> credentials)
```

Comment:
This `postProcess` method is invoked on every call to `resolver.resolve()`, at least as far as I can see, which is called immediately before the old `diffLogger.accept()` call:
https://github.com/elastic/elasticsearch/blob/e5cdf697d60222c3d3213bbfe2d7f269eb9a9610/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java#L322-L330

Can you clarify the aim of this PR? If it's just cleanup I'll give it a LGTM, but if the goal is to address @ywangd's comment [here](https://github.com/elastic/elasticsearch/pull/91066#issuecomment-1291457902) I don't think it does it.

Reply:
Hmm. Technically, I did what I set out to do in this PR description, I moved the public keys comparison to the metadata resolver.

However, I had made an assumption that SAML realm code was only running the metadata resolver periodically, and asynchronously. By moving the comparison to the resolver, I would piggyback off of any existing asynchronous scheduling of the resolver.

I am starting to think SAML realm had an issue before I made my change. 

Big picture, my goal of the previous PR and this PR was to add diagnostic info for diagnosing SAML signature verification logs. I did that, but this and any future changes to SAML realm potentially add extra load in the authentication path.

What do you think I should do with this PR? Does it make sense to merge it, and start tracking a separate fix for asynchronous execution of the resolver?

Reply:
> I am starting to think SAML realm had an issue before I made my change.

I agree, and I think that's why Yang said it would be high effort to fix. I think we should:
1) Go ahead and merge this PR; the code that calls `postprocess()` in `resolve()` is in a third-party library, so enhancements to the library could help take advantage of this (and it's possible I'm missing something). As I said, I think this works a reorganization and I think it makes sense to do it in the `postProcess` method.
2) Open an issue for the concerns Yang raised so we don't lose track of them. Maybe we can take a crack at it next tech debt week or something.